### DISCUSSION TITLE
Esp32 and YDLIDAR X3 PRO

### DISCUSSION THREAD

**Discussion by hyunjun1234** [makerspet/support#16] [Support Q&A]
How can I get data from lidar sensor. I have Esp32 dev module, YDLIDAR X3 PRO, serial port adapter board, usb c-type to ttl. Can I get angle with distance?

**Comment by makerspet** [comment]
Hello,

look at the library sample sketches. The sketches print out distance and
angle data to the Serial Monitor.

- install the LDS library in your Arduino IDE
- open the YDLIDAR X4 sample sketch from LDS library, File -> Examples ->
LDS -> sketch name
- change the #include in the sketch from YDLIDAR X4 to X3 PRO. See the list
of available include files in LDS/src folder
- change the Lidar class from X4 to X3 PRO; read the X3 PRO include file to
find the class name
- note Lidar pin names and corresponding GPIO numbers in the example sketch
- note how the sketch prints out distances and angles to Serial Monitor
- compile your sketch, make sure there are no errors
- read the X3 PRO datasheet to find the pinout for your X3 PRO here
https://github.com/kaiaai/awesome-2d-lidars
- connect your X3 PRO power to 5V 1A; connect your Lidar signal pins to
ESP32 GPIOs noted in the sample sketch
- test your sketch with your Lidar; use the Serial Monitor in the Arduino
IDE to inspect the distances and angles

**Comment by hyunjun1234** [comment]
![KakaoTalk_20241203_114916539](https://github.com/user-attachments/assets/1bf1d2da-8f96-423c-a708-b36a612f7de8)
![KakaoTalk_20241203_114916539_01](https://github.com/user-attachments/assets/62accbe2-3
[Copyright 2023-2024 REMAKE.AI, KAI.txt](https://github.com/user-attachments/files/17986648/Copyright.2023-2024.REMAKE.AI.KAI.txt)
33a-4fa3-9dec-4ede89019083)
![KakaoTalk_20241203_114916539_02](https://github.com/user-attachments/assets/cc0bca85-eb88-4941-aa46-8a64b3dd12cd)
There is no data on the serial Monitor. I connected 5V, TX(ttl) to RX(esp32 16pin), RX(ttl) to Tx(esp32 17pin). Thank you so much about your fast response. I am still expecting your response :)

**Comment by makerspet** [comment]
Don't use the serial adapter. Connect your Lidar to ESP32 pins.

The serial adapter is necessary only if you want to connect your Lidar to a
PC or Raspberry Pi/etc.

**Reply by hyunjun1234** [reply]
Thank you so much!!!!!!!!!!!!!!! 😄😄😄 감사합니다
How can I control motor speed??
then, Can I get a program that can see the 2D map..????

**Comment by makerspet** [comment]
I use ROS2 RViz for Lidar visualization, but it works with robots that have
a Lidar.

I don't have anything else. I'd suggest writing a Python script that
receives text over a COM port from the Serial Monitor and plots the
angles/distances the way you like it.

**Reply by hyunjun1234** [reply]
[208801 5330.00 7.44.txt](https://github.com/user-attachments/files/17987295/208801.5330.00.7.44.txt)
Sometime distance is 0. It is very big noise on the autonomous system. 
Is there any solution?

**Comment by makerspet** [comment]
I'd suggest googling. There are denoise algorithms out there.

**Reply by hyunjun1234** [reply]
I'll do that. Thank you!
